auth-tokens/src/main/java/com/palantir/tokens/auth/BearerToken.java

```diff
@@ -51,13 +64,26 @@
     public static BearerToken valueOf(String token) {
         AuthTokensPreconditions.checkArgument(token != null, "BearerToken cannot be null");
         AuthTokensPreconditions.checkArgument(!token.isEmpty(), "BearerToken cannot be empty");
-        if (!VALIDATION_PATTERN.matcher(token).matches()) {
+        if (!isValidBearerToken(token)) {
             log.trace("Error parsing BearerToken, must match pattern {}: {}", VALIDATION_PATTERN_STRING, token);
             throw new IllegalArgumentException("BearerToken must match pattern " + VALIDATION_PATTERN_STRING);
         }
         return ImmutableBearerToken.of(token);
     }
 
+    private static boolean isValidBearerToken(String token) {
+        boolean isAtEnd = false;
+        int length = token.length();
+        for (int i = 0; i < length; i++) {
+            char character = token.charAt(i);
+            if ((isAtEnd && character != '=') || !allowedCharacters.get(character)) {
```

Comment:
while I think this is correct, it took me a while to convince myself. i guess the tests should be sufficient to make this stable, but wondering if a more procedural algorithm (with nested if/else etc) would make this easier to follow?

Reply:
hopefully improved